pkg/kubelet/pod_sandbox_deleter.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"sort"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/klog/v2"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+)
+
+const (
+	// The number of sandboxes which can be deleted in parallel.
+	sandboxDeletionBufferLimit = 20
+)
+
+type sandboxStatusByCreatedList []*runtimeapi.PodSandboxStatus
+
+type podSandboxDeleter struct {
+	worker chan<- string
+}
+
+func (a sandboxStatusByCreatedList) Len() int      { return len(a) }
+func (a sandboxStatusByCreatedList) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a sandboxStatusByCreatedList) Less(i, j int) bool {
+	return a[i].CreatedAt > a[j].CreatedAt
+}
+
+func newPodSandboxDeleter(runtime kubecontainer.Runtime) *podSandboxDeleter {
+	buffer := make(chan string, sandboxDeletionBufferLimit)
+	go wait.Forever(func() {
+		for id := range buffer {
+			if err := runtime.DeleteSandbox(id); err != nil {
+				klog.Warningf("[pod_sandbox_deleter] DeleteSandbox returned error for (id=%v): %v", id, err)
+			}
+		}
+	}, 0)
+
+	return &podSandboxDeleter{
+		worker: buffer,
+	}
+}
+
+// deleteSandboxesInPod issues sandbox deletion requests for all inactive sandboxes after sorting by creation time
+// and skipping toKeep number of sandboxes
+func (p *podSandboxDeleter) deleteSandboxesInPod(podStatus *kubecontainer.PodStatus, toKeep int) {
+	sandboxIDs := sets.NewString()
+	for _, containerStatus := range podStatus.ContainerStatuses {
+		sandboxIDs.Insert(containerStatus.PodSandboxID)
+	}
+	sandboxStatuses := podStatus.SandboxStatuses
+	if toKeep > 0 {
+		sort.Sort(sandboxStatusByCreatedList(sandboxStatuses))
+	}
+
+	for i := len(sandboxStatuses) - 1; i >= toKeep; i-- {
+		if _, ok := sandboxIDs[sandboxStatuses[i].Id]; !ok && sandboxStatuses[i].State != runtimeapi.PodSandboxState_SANDBOX_READY {
+			select {
+			case p.worker <- sandboxStatuses[i].Id:
+			default:
+				klog.Warningf("Failed to issue the request to remove sandbox %v", sandboxStatuses[i].Id)
```

Comment:
so if there are 20 (`sandboxDeletionBufferLimit = 20`) sandboxes are being queued for deletion, this one will not be deleted and re-tried? Is it intended?

Reply:
The other pod sandbox gc routines will clean this up if this happens.  This change is to trigger sandbox cleanup quickly to make pod deletion time quicker.

Reply:
thank you! This makes sense. But than the comment https://github.com/kubernetes/kubernetes/pull/92817/files#r450545856 about postponing pruning goes against this optimization, is it?

Reply:
My other comment is saying that we could choose to run this logic _only_ during deletion, rather than when sandboxes are recreated.  The "other" gc I mention above is here: https://github.com/kubernetes/kubernetes/blob/21953d15ea48972f20a8de29d58bd5ce6d913914/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L408.  That logic will still do pruning